### PR TITLE
Support resolving MethodInfo from strongly-typed route handler delegate

### LINKED ISSUE #53171: Make it possible to resolve the method info from the specific delegate type when using AddRouteHandler
For more NativeAOT size savings, we can avoid preserving all type information for methods pointed to by delegates if we enable this. See https://github.com/dotnet/runtime/pull/96166#issuecomment-1877438742 for more details.

We essentially need to add another thunk [here](https://github.com/dotnet/aspnetcore/blob/e6be3e95fec33ca3a3b576df4b265ead680a91a0/src/Http/Routing/src/RouteEndpointDataSource.cs#L51) to allow resolution of MethodInfo from the delegate when using RDG (the request delegate generator).

cc @captainsafia

### PR DESCRIPTION
**Pull request by captainsafia**
Closes https://github.com/dotnet/aspnetcore/issues/53171.

Indentation changes are a reaction to https://github.com/dotnet/runtime/pull/92791.

### CODE REVIEW

File: src/Http/Routing/src/Builder/RouteHandlerServices.cs [hunk 1]
```diff
@@ -29,14 +29,16 @@ public static class RouteHandlerServices
...
     /// <returns></returns>
     public static RouteHandlerBuilder Map(
             IEndpointRouteBuilder endpoints,
             string pattern,
             Delegate handler,
             IEnumerable<string>? httpMethods,
             Func<MethodInfo, RequestDelegateFactoryOptions?, RequestDelegateMetadataResult> populateMetadata,
-            Func<Delegate, RequestDelegateFactoryOptions, RequestDelegateMetadataResult?, RequestDelegateResult> createRequestDelegate)
+            Func<Delegate, RequestDelegateFactoryOptions, RequestDelegateMetadataResult?, RequestDelegateResult> createRequestDelegate,
+            MethodInfo methodInfo)
```
Comment (davidfowl) [inline src/Http/Routing/src/Builder/RouteHandlerServices.cs:41]: This needs to be optional or it's a breaking change.

File: src/Http/Routing/src/RouteEndpointDataSource.cs [hunk 2]
```diff
@@ -133,7 +137,7 @@ internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
         // ApplicationBuilder.Build(). This was observed in MapSignalRTests and is not very useful. Maybe if we come up
         // with a better heuristic for what a useful method name is, we could use it for everything. Inline lambdas are
         // compiler generated methods so they are filtered out even for route handlers.
-        if (isRouteHandler && TypeHelper.TryGetNonCompilerGeneratedMethodName(handler.Method, out var methodName))
+        if (isRouteHandler && TypeHelper.TryGetNonCompilerGeneratedMethodName(entry.Method, out var methodName))
```
Comment (davidfowl) [inline src/Http/Routing/src/RouteEndpointDataSource.cs:140]: Other pieces of code needs to be updated. I would resolve the method info at the top and use it everywhere.

File: src/Http/Routing/src/RouteEndpointDataSource.cs [hunk 3]
```diff
@@ -213,11 +213,11 @@ internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
             Debug.Assert(entry.InferMetadataFunc != null, "A func to infer metadata must be provided for route handlers.");
 
             rdfOptions = CreateRdfOptions(entry, pattern, builder);
-            rdfMetadataResult = entry.InferMetadataFunc(entry.RouteHandler.Method, rdfOptions);
+            rdfMetadataResult = entry.InferMetadataFunc(methodInfo, rdfOptions);
         }
 
         // Add delegate attributes as metadata before entry-specific conventions but after group conventions.
-        var attributes = handler.Method.GetCustomAttributes();
+        var attributes = entry.Method.GetCustomAttributes();
```
Comment (davidfowl) [inline src/Http/Routing/src/RouteEndpointDataSource.cs:220]: Why is this using entry.Method and not methodInfo?

Comment (captainsafia) [reply src/Http/Routing/src/RouteEndpointDataSource.cs:220]:
```suggestion
        var attributes = methodInfo.GetCustomAttributes();
```

**Review by davidfowl — APPROVED** [review]

**Pull request merged**